app/Models/BeatmapSet.php

```diff
@@ -471,6 +472,180 @@ public function replies($time = null)
+    }
+
+    public function storage()
+    {
+        if ($this->_storage === null) {
+            $this->_storage = new StorageWithUrl();
+        }
+
+        return $this->_storage;
+    }
+
+    // todo: generalize method
+    public function oszDownloadURL($noVideo = 1)
+    {
+        $mirrors = config('osu.beatmap_processor.mirrors_to_use');
+        $mirror = BeatmapMirror::find($mirrors[array_rand($mirrors)]);
+
+        $diskFilename = $serveFilename = $this->filename;
+        $time = time();
+        $checksum = md5("{$this->beatmapset_id}{$diskFilename}{$serveFilename}{$time}{$noVideo}{$mirror->secret_key}");
+
+        $url = "{$mirror->base_url}d/{$this->beatmapset_id}?fs=".rawurlencode($serveFilename).'&fd='.rawurlencode($diskFilename)."&ts=$time&cs=$checksum&u=0&nv=$noVideo";
+
+        return $url;
+    }
+
+    public function regenerateCovers()
+    {
+        $time = time();
+
+        $tmpBase = sys_get_temp_dir()."/bm/$this->beatmapset_id";
+        $osz = "$tmpBase/$time/$this->beatmapset_id.zip";
+        $workingFolder = "$tmpBase/$time/working";
+        $outputFolder = "$tmpBase/$time/out";
+
+        // make our temp folders if they don't exist
+        if (!is_dir($workingFolder)) {
+            mkdir($workingFolder, 0755, true);
+        }
+        if (!is_dir($outputFolder)) {
+            mkdir($outputFolder, 0755, true);
+        }
+
+        // download and extract beatmap
+        $ok = copy($this->oszDownloadURL(), $osz);
+        if (!$ok) {
+            throw new BeatmapProcessorException('Error retrieving beatmap');
+        }
+        $zip = new \ZipArchive;
+        $zip->open($osz);
+        $zip->extractTo($workingFolder);
+        $zip->close();
+
+        // grab the first beatmap (as per old implementation) and scan for background image
+        $beatmap = $this->beatmaps()->first();
+        $beatmapFilename = $beatmap->filename;
+        $bg = $this::scanBMForBG("$workingFolder/$beatmapFilename");
+        $bg = str_replace('\\', '/', $bg); // windows pathing woo
+        if (!$bg) {
+            deltree($tmpBase);
+            $this->update(['cover_updated_at' => $this->freshTimestamp()]);
+
+            return true;
+        }
+
+        $bg_file = ci_file_search("{$workingFolder}/{$bg}");
+        if (!$bg_file) {
+            deltree($tmpBase);
+            throw new BeatmapProcessorException("Background image missing: {$bg}");
+        }
+
+        // upload original image
+        $this->storage()->put("/beatmaps/{$this->beatmapset_id}/covers/raw.jpg", file_get_contents($bg_file));
+        $originalImage = preg_replace("/https?:\/\//", '', $this->coverImageURL('raw'));
+
+        // upload optimized version
+        $resizerEndpoint = config('osu.beatmap_processor.thumbnailer');
+        $optimizedImage = preg_replace("/https?:\/\//", '', $this->coverImageURL('fullsize'));
+
+        $ok = copy("$resizerEndpoint/optim/$originalImage", "$outputFolder/fullsize.jpg");
```

Comment:
this whole thing (and the one in loop below which does similar thing) probably can be moved into its own method.